File: ImperatorToCK3/Source/Imperator/Provinces/PopFactory.cpp
#include "PopFactory.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"



Imperator::Pop::Factory::Factory() {
	registerKeyword("type", [this](std::istream& theStream) {
		pop->type = commonItems::getString(theStream);
	});
	registerKeyword("culture", [this](std::istream& theStream) {
		pop->culture = commonItems::getString(theStream);
	});
	registerKeyword("religion", [this](std::istream& theStream) {
		pop->religion = commonItems::getString(theStream);
	});
	registerMatcher(commonItems::catchallRegexMatch, commonItems::ignoreItem);
}


std::unique_ptr<Imperator::Pop> Imperator::Pop::Factory::getPop(const std::string& idString, std::istream& theStream) {
	pop = std::make_unique<Pop>();
	pop->ID = std::stoull(idString);

	parseStream(theStream);

	return std::move(pop);
}

File: ImperatorToCK3/Imperator/Pops/PopFactory.cs
using commonItems;

namespace ImperatorToCK3.Imperator.Pops;

public partial class Pop {
	static Pop() {
		popParser.RegisterKeyword("type", reader => tempPop.Type = reader.GetString());
		popParser.RegisterKeyword("culture", reader => tempPop.Culture = reader.GetString());
		popParser.RegisterKeyword("religion", reader => tempPop.Religion = reader.GetString());
		popParser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreAndLogItem);
	}
	public static Pop Parse(string idString, BufferedReader reader) {
		tempPop = new Pop(ulong.Parse(idString));
		popParser.ParseStream(reader);
		return tempPop;
	}

	private static Pop tempPop = new(0);
	private static readonly Parser popParser = new();
}